File: examples/node/03_serve_client.mjs
/** HTTP client against lucy serve — BYO frontend pattern. */
import { readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { spawn } from "node:child_process";
import { createLucyClient, resolveLucyBinary } from "../../js/src/index.js";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
const req = JSON.parse(await readFile(path.join(root, "examples/shared/samples.json"), "utf8"));
const bin = await resolveLucyBinary();
const addr = "127.0.0.1:17490";
const child = spawn(bin, ["serve", addr], { stdio: ["ignore", "ignore", "pipe"] });
try {
  await new Promise((r) => setTimeout(r, 500));
  const c = createLucyClient(`http://${addr}`);
  console.log("version", await c.version());
  console.log("floors keep", (await c.floors()).keep_floor);
  const { board } = await c.buildLPD(req.samples, req.options);
  console.log("top", board.top[0].id);
  const pdf = await c.sitePDF(req.samples, req.options);
  const csv = await c.csv(req.samples, req.options);
  console.log("pdf bytes", pdf.length, "csv lines", csv.trim().split("\n").length);
  console.log("OK serve client");
} finally {
  child.kill("SIGTERM");
}

File: examples/python/03_serve_client.py
#!/usr/bin/env python3
"""LucyClient against lucy serve."""
from __future__ import annotations

import json
import subprocess
import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "python" / "src"))

from lucy import LucyClient
from lucy._bin import default_binary

req = json.loads((ROOT / "examples/shared/samples.json").read_text())
addr = "127.0.0.1:17491"
proc = subprocess.Popen(
    [str(default_binary()), "serve", addr],
    stdout=subprocess.DEVNULL,
    stderr=subprocess.DEVNULL,
)
try:
    time.sleep(0.5)
    c = LucyClient(f"http://{addr}")
    print("version", c.version())
    print("floors keep", c.floors()["keep_floor"])
    board = c.build_lpd(req["samples"], req.get("options"))["board"]
    print("top", board["top"][0]["id"])
    pdf = c.site_pdf(req["samples"], req.get("options"))
    csv = c.csv(req["samples"], req.get("options"))
    print("pdf", len(pdf), "csv lines", len(csv.strip().splitlines()))
    print("OK python serve client")
finally:
    proc.terminate()
    proc.wait(timeout=5)
